### ISSUE TITLE
Restore Subnet Preference Logic

### ISSUE THREAD

**Issue by cooperlees** [chihaya/chihaya#267]
Can we please re-add the Subnet preference logic that we added originally back in 2014.

Summary:
- Config file allows a CIDR subnet size for v4 and v6
- We store peers and seeders in subnet addressed maps
- Prefer in subnet seeders, peers then return other randoms to the requested NUM_WANTED (or what ever the correct name is).

What's the preferred way to re-implement this today?

**Comment by jzelinskie** [comment]
I'm not sure that all trackers are going to want the overhead this behavior would require. Before working on this, we should determine if we can create the feature without compromising the performance for users with no desire to enable this feature. If there is no reasonable compromise (worst case scenario), an alternative storage implementation could be made and maintained outside the core project and we can ensure the API stays consistent.

---

Let's evaluate the pros and cons of the previous system (peers additionally sharded by a pre-configured subnet):

### Pros

* Fast for the FB use case.

### Cons

* Semi-obscure feature. I can't see many people using it without a carefully planned (overlay) network.
* In the old implementation users not using the feature had one level of indirection and some extra memory allocations.
  * We could add a more code to avoid the indirection and allocations when not in use.
  * This could also be a pro.
* If data ever persisted, reconfiguration would require data migration.
  * This is true of sharding in general and our current storage is no exception.

---

I yield now to additional implementation ideas. I know @mrd0ll4r has had a few thoughts in the past about something trie-like that could index the peers in a swarm.

**Comment by mrd0ll4r** [comment]
In my opinion we should not alter the `memory` implementation we have now. It's fast, non-persistent and it makes the tracker behave the way it should. The implementation we had in v1 was memory-hungry, much slower and, as @jzelinskie pointed out, rather difficult to understand.

We should instead write an implementation of the `storage.PeerStore` interface that behaves the way facebook needs it to behave - after all that's why we made it modular. I don't have a strong opinion about where this should take place. If we do it in a repository under the chihaya organization though, we can keep it up to date with eventual API changes.

## Possible implementations

### Trie
A while ago I wrote `netmatch`, which is essentially a prefix-tree for IP addresses. This got us thinking: We could use the same structure to store peers. We could then answer an announce by depth-first search in that tree, which would return peers with a numerically close IP address.

**Pros**
- Prefix length is implicit - will always return the closest peers.
- Trivial to implement, easy to understand.
- Might be useful to the general public.

**Cons**
- [Trees in go are painfully slow](http://benchmarksgame.alioth.debian.org/u64q/performance.php?test=binarytrees). This is due to all the pointer handling and GC, I believe. It'd be possible to reduce the tree in height by making it *256*-ary, which might improve performance. It's also possible to flatten it out altogether (storing the peers as a sorted array, basically), but this complicates the DFS.
- It does not behave the way facebook needs it to:  
    You want all peers from the same subnet, then random other peers, until you have `num_want`. This makes sense for your infrastructure, as most traffic will be intra-rack, but there's also data flowing between random racks. This does for a fast, parallel deployment.  
    However, using a trie would return peers from the same subnet, then peers from a slightly larger subnet the announcing peer is still a part of. This would, correct me if I'm wrong, have data flowing from one rack to the racks immediately next to it. Data traffic would be more like a ring, the deployment would be less parallel.

### Map-based (sharding by prefix)
This is what we did in v1. It's proven to work, although it wasn't very fast and consumed a lot of memory.  
Be careful with putting a lot of maps in a lot of maps, i.e. something like every infohash gets a map of seeders and leechers, each with a map of prefixes to peers (just an example). Maps have a substantial memory footprint. Another mistake we made in v1 was to store things twice, most notable we mapped something like a `peerKey` (peer ID + IP/Port) to a `Peer`, which contained, big surprise, an `ID`, `IP` and `Port` field.

About performance: We have a benchmark suite that benchmarks any implementation of the `PeerStore` interface. We should be able to optimize much better now. :)
About memory: I wrote a tool to test the amount of memory consumed per peer/per infohash, I'd be glad to help bring down the memory footprint.

**Pros**
- It does what you need.
- It works.
- We basically have it implemented already, just need to make it conform to the new interface, and optimize.
- It probably works for other cases with similar requirements.

**Cons**
- It might be unfixably slow or memory-hungry, we'll have to see.


In conclusion, I'm in favor of implementing the map-based solution.

**Comment by jzelinskie** [comment]
I made a quick pass at a minimal modification of our memory storage to support this behavior.
[Check out the branch](https://github.com/jzelinskie/chihaya/tree/memorybysubnet/storage/memorybysubnet). The tests pass, so you should be able to run and compare the benchmarks, too.
